alembic/versions/e372e4daf771_add_many_to_many_tables_for_stripe_.py

```diff
@@ -0,0 +1,322 @@
+            .format(model, id, e)
+
+
+def add_model_by_txn(txn, multi=False):
+    # Adds a model according to the fk_model stored on a transaction, and returns its name
+    connection = op.get_bind()
+
+    share, error = (get_model_share(txn.fk_id, txn.fk_model)) if multi else (int(txn.amount), None)
+
+    if error:
+        return '', error
+
+    if txn.fk_model == "Attendee":
+        [attendee] = connection.execute(
+            attendee_table.select().where(
+                attendee_table.c.id == txn.fk_id
+            )
+        )
+
+        if txn.amount <= (attendee.amount_paid * 100) or multi:
+            connection.execute(
+                attendee_txn_table.insert().values({
+                    'id': str(uuid.uuid4()),
+                    'txn_id': txn.id,
+                    'attendee_id': attendee.id,
+                    'share': share
+                })
+            )
+
+            return attendee.first_name + " " + attendee.last_name, None
+
+    elif txn.fk_model == "Group":
+        [group] = connection.execute(
+            group_table.select().where(
+                group_table.c.id == txn.fk_id
+            )
+        )
+        if txn.amount <= (group.amount_paid * 100) or multi:
+            connection.execute(
+                group_txn_table.insert().values({
+                    'id': str(uuid.uuid4()),
+                    'txn_id': txn.id,
+                    'group_id': group.id,
+                    'share': share
+                })
+            )
+
+            return group.name, None
+
+def upgrade():
+    op.create_table('stripe_transaction_group',
+    sa.Column('id', residue.UUID(), nullable=False),
+    sa.Column('txn_id', residue.UUID(), nullable=False),
+    sa.Column('group_id', residue.UUID(), nullable=False),
+    sa.Column('share', sa.Integer(), nullable=False),
+    sa.ForeignKeyConstraint(['group_id'], ['group.id'], name=op.f('fk_stripe_transaction_group_group_id_group')),
+    sa.ForeignKeyConstraint(['txn_id'], ['stripe_transaction.id'], name=op.f('fk_stripe_transaction_group_txn_id_stripe_transaction')),
+    sa.PrimaryKeyConstraint('id', name=op.f('pk_stripe_transaction_group'))
+    )
+    op.create_table('stripe_transaction_attendee',
+    sa.Column('id', residue.UUID(), nullable=False),
+    sa.Column('txn_id', residue.UUID(), nullable=False),
+    sa.Column('attendee_id', residue.UUID(), nullable=False),
+    sa.Column('share', sa.Integer(), nullable=False),
+    sa.ForeignKeyConstraint(['attendee_id'], ['attendee.id'], name=op.f('fk_stripe_transaction_attendee_attendee_id_attendee')),
+    sa.ForeignKeyConstraint(['txn_id'], ['stripe_transaction.id'], name=op.f('fk_stripe_transaction_attendee_txn_id_stripe_transaction')),
+    sa.PrimaryKeyConstraint('id', name=op.f('pk_stripe_transaction_attendee'))
+    )
+
+    connection = op.get_bind()
+
+    txns = connection.execute(
+        stripe_txn_table.select()
+    )
+
+    errors = []
+
+    for txn in txns:
```

Comment:
This migration hinges on one very important fact: multi-model Stripe transactions ONLY occur when a model is first being created. That means, rather than faffing around and making guesses, we just need to pull the model's `amount_paid` from the Tracking object for its creation. The only major failure point left, then, is that we have to find Attendees and Groups by their name (pulled from the Stripe Transaction's description, the only record of who is in the transaction). This migration gets around that by checking the number of groups and attendees matching a name against the number of times the name occurs in the transaction's description -- if there's a mismatch, it throws an error rather than attempting to create tables for those attendees and groups.

Currently, the migration is in 'test' mode -- it won't commit changes if it detects an error, and downgrading simply destroys the new tables (which the upgrade created, so we're not losing data). This allowed easy testing with my local data, but we should still back up the server before trying this just in case.

Reply:
Turns out we weren't doing pre-schema-migration database backups for the new deployments 😬

This pull request adds a database backup step before the migration is run: https://github.com/magfest/infrastructure/pull/8

So worse comes to worst, we can shut down the system and restore from backup if the migration breaks everything!